src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -38,12 +38,14 @@ public static IList<Diagnostic> GetCompileTimeConstantViolation(SyntaxBase expre
         /// <returns>Returns true if values of the specified source type are assignable to the target type. Returns false otherwise or null if assignability cannot be determined.</returns>
         public static bool? AreTypesAssignable(TypeSymbol? sourceType, TypeSymbol? targetType)
         {
-            if (sourceType == null || targetType == null || sourceType is ErrorTypeSymbol || targetType is ErrorTypeSymbol)
+            // return null when sourceType, targetType is either null or sourceType is UnassignableTypeSymbol and contains errors -> GetDiagnostics() returns a list of errors or when targetType is UnassignableTypeSymbol
+            if (sourceType == null || targetType == null || ( sourceType is UnassignableTypeSymbol unassignableTypeSymbol && unassignableTypeSymbol.GetDiagnostics().Any() ) || targetType is UnassignableTypeSymbol)
             {
                 return null;
             }
 
-            if (sourceType is AnyType)
+            // sourceType is either AnyType or is UnassignableTypeSymbol and does not contain any errors.
+            if (sourceType is AnyType || (sourceType is UnassignableTypeSymbol _unassignableTypeSymbol && !_unassignableTypeSymbol.GetDiagnostics().Any()))
```

Comment:
Same comment here - why would we ever want to allow assignment of the type named `UnassignableTypeSymbol`? Sounds like perhaps we need to rename it if there are some legitimate cases where assignment is allowed :)

Reply:
Yeah, removed the update and left it as is was before.